### ISSUE TITLE
Add rotation

### ISSUE THREAD

**Issue by vjeux** [excalidraw/excalidraw#15]
Zwibbler allows you to rotate all the elements. I'm not sure I've ever used it and it's been more annoying in my experience. Creating an issue to think about it

- [x] Rotate any element
- [x] Lock on 0, 45, 90, 135, 180, etc

**Comment by Fausto95** [comment]
In my opinion I’d only implement this for the arrow since I’m not sure that rotating other shapes would be useful

**Comment by vjeux** [comment]
Arrow don't need to be rotated (in what I had in mind for this PR), changing the two end points would work just fine (basically resizing).

**Comment by dwelle** [comment]
Rotating other shapes, namely rectangle, can be useful to create diamond shapes for conditional logic diagrams.

**Comment by vjeux** [comment]
We could add support for diamond directly as well

**Comment by szympajka** [comment]
Would be nice to rotate shapes :)

**Comment by GrimTheReaper** [comment]
There are some shapes that would be nice to have rotated, adding shapes instead of adding a simple rotate feature seems off.

**Comment by vjeux** [comment]
@GrimTheReaper do you have concrete examples?

**Comment by GrimTheReaper** [comment]
@vjeux concrete examples of what? Implementing a rotate feature or the fact that most shape editors have a rotate feature? Also on another note, can you even make the excalidraw logo in excalidraw?

**Comment by vjeux** [comment]
My mental process for building product is to ask why people want a certain feature, in many cases there's actually a better solution to solve their problem. For example if you want to make a diamond, it's convoluted to draw a square and then rotate it.

It doesn't mean that we shouldn't implement rotate, I'm just curious what use cases people have with it. It's not a trivial feature to implement so if we can get away without it that would be nice :)

**Comment by GrimTheReaper** [comment]
What if, I wanted to have an offset diamond? inbetween 45 and 90

**Comment by vjeux** [comment]
@GrimTheReaper yeah, you'd need rotation for it. Is this something you actually wanted to do in practice?

**Comment by GrimTheReaper** [comment]
@vjeux realistically, probably not that specific example. But I feel that a better rational for adding it is the fact that without it, shapes can be more limited in terms of being on the canvas. While I fail to better articulate this fact, I feel that adding it would be substantial while being a little more trivial.

**Comment by dwelle** [comment]
Let's keep this open for now until we polish up what we have so far, or until someone comes here with a dire need :).

**Comment by vjeux** [comment]
Just to be clear, if someone implements it that would be a welcome addition. But it’s not at a very high priority.

**Comment by GrimTheReaper** [comment]
Fine by me, I just wanted to make sure it would be on a roadmap somewhere.

**Comment by xenophilicibex** [comment]
I think it might be nice to have preset rotations for each shape.

So rectangles, diamonds, ellipses, and lines don't need them. 

But as soon as they're more polygons like triangles (△, ▹, ◸, +) and hexagons (⬡,⎔), rotation becomes the easiest way to configure it.

**Comment by RichiCoder1** [comment]
^ Same need. I'd like to align text with some arrows that aren't flat basically.

**Comment by jdestgermain** [comment]
Text rotation + rectangle rotation is something that I've run into wanting so just wanted to add my use case.

I am not sure what industry or what purpose everyone is using the tool for but for me I stumbled across it from [HN ](https://news.ycombinator.com/item?id=22101381) and so far have liked using it for Mechanical Engineering purposes. Specifically, I really like it for brainstorming and quick drawings of [Free Body Diagrams](https://en.wikipedia.org/wiki/Free_body_diagram).  

So for instance, the drawing in wikipedia for the FBD page **can** be done with a diamond if you only want it to be a square and at a perfect 45 degree ramp (as shown), but this is idealistic.  Any other example becomes harder to illustrate.

![FBD example](https://upload.wikimedia.org/wikipedia/commons/thumb/e/e5/Free_body_diagram2.svg/200px-Free_body_diagram2.svg.png)

Here it is, drawn in Excalibur (notice how I had to do a 45 degree incline plane, no text rotated, etc.)
![image](https://user-images.githubusercontent.com/11359044/77551581-871e4a80-6e80-11ea-9842-72cb618b725e.png)
